### PR TITLE
Expand and fix testing of inbound filter chains

### PR DESCRIPTION
**Pull request by howardjohn**
This expands a bunch of testing of inbound filter chains, in particular with mtls modes. Part of this is fixing the tests to properly account for the disabled filter predicate, which is why some tests results have changed to closer match real envoy behavior. Additionally, there are a lot of incorrect behaviors that are marked

### CODE REVIEW

File: pilot/pkg/simulation/traffic.go [hunk 1]
```diff
@@ -236,29 +266,49 @@ func (sim *Simulation) Run(input Call) (result Result) {
 	}
 	result.ListenerMatched = l.Name
 
-	// Apply listener filters. This will likely need the TLS inspector in the future as well
-	if _, f := xdstest.ExtractListenerFilters(l)[xdsfilters.HTTPInspector.Name]; f {
+	hasTLSInspector := hasFilterOnPort(l, xdsfilters.TLSInspector.Name, input.Port)
```
Comment (howardjohn) [inline pilot/pkg/simulation/traffic.go:269]: fix to respect the disable predicate

File: pilot/pkg/simulation/traffic.go [hunk 2]
```diff
@@ -236,29 +266,49 @@ func (sim *Simulation) Run(input Call) (result Result) {
...
 		return
 	}
 	result.FilterChainMatched = fc.Name
+	// Plaintext to TLS is an error
 	if fc.TransportSocket != nil && input.TLS == Plaintext {
 		result.Error = ErrTLSError
 		return
 	}
+	// mTLS listener will only accept mTLS traffic
+	if fc.TransportSocket != nil && sim.requiresMTLS(fc) != (input.TLS == MTLS) {
```
Comment (howardjohn) [inline pilot/pkg/simulation/traffic.go:295]: actually assert that mtls to tls, and tls to mtls is an error

File: pilot/pkg/simulation/traffic.go [hunk 3]
```diff
@@ -488,6 +554,28 @@ func filter(chains []*listener.FilterChain,
 	return res
 }
 
+func protocolToMTLSAlpn(s Protocol) string {
```
Comment (howardjohn) [inline pilot/pkg/simulation/traffic.go:557]: setup defaults so we don't need them in every test

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 4]
```diff
@@ -722,3 +722,28 @@ func buildOutboundCatchAllNetworkFilterChains(_ *ConfigGeneratorImpl,
 	chains = append(chains, &listener.FilterChain{Name: VirtualOutboundCatchAllTCPFilterChainName, Filters: filterStack})
 	return chains
 }
+
+// EvaluateListenerFilterPredicatesInternal runs through the ListenerFilterChainMatchPredicate logic
+// This is expose for testing only, and should not be used in XDS generation code
```
Comment (tariq1890) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:727]:
```suggestion
// This is exposed for testing only, and should not be used in XDS generation code
```

File: pilot/pkg/simulation/traffic.go [hunk 5]
```diff
@@ -295,6 +345,22 @@ func (sim *Simulation) Run(input Call) (result Result) {
...
+	}
+	t := &tls.DownstreamTlsContext{}
+	if err := ptypes.UnmarshalAny(fc.GetTransportSocket().GetTypedConfig(), t); err != nil {
+		sim.t.Fatal(err)
+	}
+
+	if len(t.GetCommonTlsContext().GetTlsCertificateSdsSecretConfigs()) == 0 {
+		return false
+	}
+	// This is a lazy heuristic, we could check for explicit default resource or spiffee if it becomes neccesary
```
Comment (tariq1890) [inline pilot/pkg/simulation/traffic.go:360]:
```suggestion
	// This is a lazy heuristic, we could check for explicit default resource or spiffe if it becomes necessary
```

File: pilot/pkg/simulation/traffic.go [hunk 6]
```diff
@@ -147,16 +156,20 @@ type Result struct {
 	RouteConfigMatched string
 	VirtualHostMatched string
 	ClusterMatched     string
-	// StrictMatch controls whether we will strictly match the result. If not set,
-	// empty fields will be ignored, allowing testing only fields we care about This
-	// allows asserting that the result is *exactly* equal, allowing asserting a
-	// field is empty
+	// StrictMatch controls whether we will strictly match the result. If not set, empty fields will
```
Comment (tariq1890) [inline pilot/pkg/simulation/traffic.go:159]:
```suggestion
	// StrictMatch controls whether we will strictly match the result. If  unset, empty fields will
```

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go (hunk 4 above)
Comment (tariq1890) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:727]: EvaluateListenerFilterPredicatesInternal should be renamed IMO. Honestly specifying `Internal` in the name is redundant. Go's treatment of camelCase and PascalCase method names already imply this

File: pilot/pkg/simulation/traffic.go [hunk 7]
```diff
@@ -147,16 +156,20 @@ type Result struct {
...
+	// be ignored, allowing testing only fields we care about This allows asserting that the result
+	// is *exactly* equal, allowing asserting a field is empty
 	StrictMatch bool
-	t           test.Failer
+	// If set, this will mark a test as skipped. Note the result is still checked first - we skip only
+	// if we pass the test. This is to ensure that if the behaviour changes, we still capture it; the skip
```
Comment (tariq1890) [inline pilot/pkg/simulation/traffic.go:164]: Apparently, `behaviour` is a misspelt word. Sadness

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:727]: Its used by other packages, so cannot be camelCase

Comment (tariq1890) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:727]: Yeah, I see that. What I meant to say is that `EvaluateListenerFilterPredicates` should be chosen over `EvaluateListenerFilterPredicatesInternal`

File: pilot/pkg/networking/core/v1alpha3/listener_builder.go [hunk 8]
```diff
@@ -722,3 +722,28 @@ func buildOutboundCatchAllNetworkFilterChains(_ *ConfigGeneratorImpl,
 	chains = append(chains, &listener.FilterChain{Name: VirtualOutboundCatchAllTCPFilterChainName, Filters: filterStack})
 	return chains
 }
+
+// EvaluateListenerFilterPredicates runs through the ListenerFilterChainMatchPredicate logic
+// This is exposed for testing only, and should not be used in XDS generation code
+func EvaluateListenerFilterPredicates(predicate *listener.ListenerFilterChainMatchPredicate, invertMatch bool, port int) bool {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener_builder.go:728]: If this is only for test, can you move it out to the test file?

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:728]: No, it's shared by multiple packages

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:728]: oh I guess we can move it to the xdstest package though

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:728]: That's good

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener_builder.go:728]: moved it

**Review by costinm — APPROVED** [review]

**Pull request merged**